go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -290,11 +287,33 @@ func (c *Controller) syncServices(key string) error {
 			// get the endpoints associated to the vip
 			eps := getLbEndpoints(endpointSlices, svcPort, family)
 			// Reconcile OVN, update the load balancer with current endpoints
-			if c.needsOVNLBUpdate(eps, service) {
-				if err := loadbalancer.UpdateLoadBalancer(lbID, vip, eps); err != nil {
-					c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToUpdateOVNLoadBalancer",
-						"Error trying to update OVN LoadBalancer for Service %s/%s: %v", name, namespace, err)
-					return err
+			if c.needsOVNLBUpdate(eps.IPs.List(), service) {
+				// If any of the lbEps contain the a host IP we add to worker/GR LB separately, and not to cluster LB
+				if hasHostEndpoints(eps.IPs.List()) && config.Gateway.Mode == config.GatewayModeShared {
+					if err := createPerNodeVIPs([]string{service.Spec.ClusterIP}, svcPort.Protocol, svcPort.Port, eps.IPs.List(), eps.Port); err != nil {
+						c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToUpdateOVNLoadBalancer",
+							"Error trying to update OVN LoadBalancer for Service %s/%s: %v", name, namespace, err)
+						return err
+					}
+					// Need to ensure that if vip exists on cluster LB we remove it
+					// This can happen if endpoints originally had cluster only ips but now have host ips
+					vip := util.JoinHostPortInt32(service.Spec.ClusterIP, svcPort.Port)
```

Comment:
you are overwriting the VIP here, bear in mind that dual-stack services have multiple vips.
The right vip processed in this iteration is created in L285
https://github.com/ovn-org/ovn-kubernetes/pull/2055/commits/0f6bddd13cf1970e2ba4e9dd79d35a73dd7ae499#diff-b61468c0028d28eb0e1060d24f53f28b6482acfbfd0cf5c15f9c23c3fd8c9f91R285